lib/puppet/application/lookup.rb

```diff
@@ -365,7 +353,30 @@ def generate_scope
       unless given_facts.instance_of?(Hash)
         raise _("Incorrect formatted data in %{fact_file} given via the --facts flag") % { fact_file: fact_file }
       end
-      node.add_extra_facts(given_facts)
+    end
+
+    unless node.is_a?(Puppet::Node) # to allow unit tests to pass a node instance
+      extra_facts = Puppet::Node::Facts.indirection.find(node, :environment => Puppet.lookup(:current_environment))
+      if extra_facts && given_facts
+        extra_facts.add_extra_values(given_facts)
+      elsif given_facts && extra_facts.nil?
+        extra_facts = Puppet::Node::Facts.new(node, given_facts)
+      elsif extra_facts.nil? && given_facts.nil?
+        raise _("No facts available for target node: %{node}") % { node: node}
+      end
```

Comment:
@joshcooper I think if the node terminus is `exec` (ENC), it's valid for `puppet lookup` to be executed without facts, since the ENC should be responsible for collecting facts/classification.

Based on this, would it make sense to raise unless the terminus is `exec`? Are there any other termini that behave similarly and we should take into consideration?

Reply:
Every node terminus kind of "does it's own thing" with respect to facts. For example, the classifier node terminus will call out to the facts terminus when [facts are not sent with the node request](https://github.com/puppetlabs/classifier/blob/6e007c5760896c0beb6cf31cf6768e729f02370c/puppet/lib/puppet/indirector/node/classifier.rb#L52). And like you said, an `exec` based node terminus may not collect any facts. I think it'd be good for the lookup application to do always call the facts indirection directly and pass those facts with the node request, for example something like:

```ruby
facts = Puppet::Node::Facts.indirection.find(node, environment: ...)
# merge extra facts
node = Puppet::Node.indirection.find(node, facts: facts, environment: ...)
```

If the facts terminus returns nil and there are extra facts, then we should use the extra facts.

If the facts terminus returns nil and no extra facts were specified, then I think that should be an error. 

If there are facts and extra facts, then I think we want to merge them and have extra facts take precedence.